src/python/nnfusion/runtime.py

```diff
@@ -1,71 +1,140 @@
-                        "-fkernel_tuning_steps=" + str(self.tuning_step),
-                        "-fir_based_fusion=1",
-                        "-fkernel_fusion_level=0",
-                        # "-fantares_mode=1",
-                        # f"-fantares_codegen_server={self.codegen_server}",
-                        "-fblockfusion_level=0"]),
-                    self.workdir)
-            modify_nnfusion_rt(rt_dir)
-            build(rt_dir)
-        if rebuild:
-            build(rt_dir)
-        self.executor = Executor(rt_dir)
-        self.input_name = self.executor.get_inputs()[0].name
-        self.output_name = self.executor.get_outputs()[0].name
-
-    def run(self, input, output):
-        in_dict = {}
-        out_dict = {}
-        if isinstance(input, list):
-            # assert len(self.executor.get_inputs()) == len(input)
-            for i in range(len(input)):
-                in_dict[self.executor.get_inputs(
-                )[i].name] = cast_pytorch_tensor(input[i])
-        else:
-            in_dict = {self.input_name: cast_pytorch_tensor(input)}
-        if isinstance(output, list):
-            # assert len(self.executor.get_outputs()) == len(output)
-            for i in range(len(output)):
-                out_dict[self.executor.get_outputs(
-                )[i].name] = cast_pytorch_tensor(output[i])
-        else:
-            out_dict = {self.output_name: cast_pytorch_tensor(output)}
 
+        def check_if_need_build():
+            if not os.path.exists(self.onnx_path):
+                return True
+
+            # Compare onnx file to check if modified
+            with tempfile.TemporaryDirectory(dir=self.workdir) as tmp:
+                temp_onnx_path = os.path.join(tmp, "temp.onnx")
+                export_onnx(temp_onnx_path)
+
+                if not filecmp.cmp(temp_onnx_path, self.onnx_path):
+                    # Replace the original to avoid exporting onnx twice
+                    os.remove(self.onnx_path)
+                    os.link(temp_onnx_path, self.onnx_path)
+                    return True
+            return False
+
+        def do_compile():
+            if not os.path.exists(self.onnx_path):
+                export_onnx(self.onnx_path)
+
+            codegen(self.onnx_path, self.compile_flag, self.workdir)
+            modify_nnfusion_rt(self.rt_dir)
+            build(self.rt_dir)
+
+        if force_build and os.path.exists(self.onnx_path):
+            os.remove(self.onnx_path)
+
+        if check_if_need_build():
+            do_compile()
+
+        self._init_executor(self.rt_dir, device=inputs[0].device)
+
+    def _init_executor(self, rt_dir, device):
+
+        libnnf_path = find_nnf_rt(rt_dir)
+        if not libnnf_path:
+            raise Exception("nnf_rt lib not found in folder {}".format(rt_dir))
+        libnnf = ctypes.cdll.LoadLibrary(libnnf_path)
+
+        ws_size = getattr(libnnf, 'get_workspace_size', None)()
+        self._reserved_mem = torch.empty(ws_size//8,
```

Comment:
round up ws_size//8?

Reply:
Noted

Reply:
Hi, it just occurred to me that we don't need to devide here as the unit is byte and tensor type is `int8`. Is that the case?

Reply:
yes